File: src/main/kotlin/com/full/stack/assignment/f1/mapper/RaceMapper.kt
package com.full.stack.assignment.f1.mapper

import com.full.stack.assignment.f1.data.cache.entity.DriverEntity
import com.full.stack.assignment.f1.data.cache.entity.RaceEntity
import com.full.stack.assignment.f1.data.cache.entity.SeasonEntity
import com.full.stack.assignment.f1.model.Race
import org.springframework.stereotype.Component

@Component
class RaceMapper(
    private val circuitMapper: CircuitMapper,
    private val driverMapper: DriverMapper,
    private val constructorMapper: ConstructorMapper
): Mapper<Race, RaceEntity> {
    override fun toDomain(entity: RaceEntity): Race {
        return Race(
            seasonYear = entity.seasonYear,
            round = entity.round,
            raceName = entity.raceName,
            circuit = circuitMapper.toDomain(entity.circuit),
            date = entity.date,
            winningDriver = driverMapper.toDomain(entity.winningDriver),
            winningConstructor = constructorMapper.toDomain(entity.winningConstructor)
        )
    }

    override fun toEntity(domain: Race): RaceEntity {
        return RaceEntity(
            seasonYear = domain.seasonYear,
            season = SeasonEntity(year = domain.seasonYear, champion = DriverEntity()),
            round = domain.round,
            raceName = domain.raceName,
            circuit = circuitMapper.toEntity(domain.circuit),
            date = domain.date,
            winningDriver = driverMapper.toEntity(domain.winningDriver),
            winningConstructor = constructorMapper.toEntity(domain.winningConstructor)
        )
    }
}

File: frontend/lib/data/remote/mappers/race_mapper.dart
import 'package:f1_api_client/api.dart';
import '../../../domain/models/race/race.dart';
import '../../../helpers/mapper/mapper.dart';
import 'driver_mapper.dart';
import 'circuit_mapper.dart';
import 'constructor_mapper.dart';

class RaceMapper extends Mapper<RaceDTO, Race, void> {
  final DriverMapper _driverMapper;
  final CircuitMapper _circuitMapper;
  final ConstructorMapper _constructorMapper;

  RaceMapper(this._driverMapper, this._circuitMapper, this._constructorMapper);

  @override
  Race map(RaceDTO param, {void metadata}) {
    final winningDriver = param.winningDriver;
    final circuit = param.circuit;
    final constructor = param.winningConstructor;

    return Race(
      year: param.seasonYear,
      round: param.round,
      name: param.raceName,
      date: param.date,
      winner: _driverMapper.map(winningDriver),
      circuit: _circuitMapper.map(circuit),
      constructor: _constructorMapper.map(constructor),
    );
  }
}
